native/native_main.py

```diff
@@ -118,6 +130,197 @@ def sanitizeFilename(fn):
+        #
+
+        #
+        # 'schtasks.exe' is limited as in it doesn't support
+        # task-time with granularity in seconds. So, falling back
+        # to PowerShell as the last resort.
+        #
+
+        # out_str = ""
+        # task_time = time.strftime("%H:%M",
+        #                           time.localtime(
+        #                               time.time() + 60))
+        #
+        # out_str = subprocess.check_output(
+        #     ["schtasks.exe",
+        #      "/Create",
+        #      "/F",
+        #      "/SC",
+        #      "ONCE",
+        #      "/TN",
+        #      "tridactyl",
+        #      "/TR",
+        #      "calc",
+        #      "/IT",
+        #      "/ST",
+        #      task_time],
+        #     shell=True)
+        # }}}
+
+        ff_bin_path = "\"%s\"" % shutil.which(ff_bin_name)
+        new_tab_url = "https://www.mozilla.org/en-US/firefox/"
+
+        if profile_dir == "auto":
+            ff_lock_path = ff_bin_path
+            ff_args = "\"%s\" \"%s\" \"%s\" \"%s\"" % \
+                ("-foreground",
+                 "-ProfileManager",
+                 "-new-tab",
+                 new_tab_url)
+        else:
+            ff_lock_path = "\"%s/%s\"" % (profile_dir,
+                                          ff_lock_name)
+            ff_args = "\"%s\" \"%s\" \"%s\" \"%s\" \"%s\"" % \
+                ("-foreground",
+                 "-profile",
+                 profile_dir,
+                 "-new-tab",
+                 new_tab_url)
+
+        try:
+            restart_ps1_content = '''$profileDir = "%s"
+if ($profileDir -ne "auto") {
+    $lockFilePath = %s
+    $locked = $true
+    $num_try = 10
+} else {
+    $locked = $false
+}
+while (($locked -eq $true) -and ($num_try -gt 0)) {
+try {
+    [IO.File]::OpenWrite($lockFilePath).close()
+    $locked=$fals
+} catch{
+    $num_try-=1
+    Write-Host "[+] Trial: $num_try [lock == true]"
+    Start-Sleep -Seconds 1
+}
+}
+if ($locked -eq $true) {
+Add-Type -AssemblyName System.Windows.Forms
+[System.Windows.MessageBox]::Show(
+    "Restarting Firefox failed. Please manually restart.",
+    "Tridactyl")
+} else {
+Write-Host "[+] Restarting Firefox ..."
+& %s %s
+}
+''' % (profile_dir, ff_lock_path, ff_bin_path, ff_args)
+
+            delay_sec = 1
+            task_cmd = "powershell"
+            task_name = "firefox-restart"
+            native_messenger_dirname = ".tridactyl"
+
+            restart_ps1_path = "%s\\%s\\%s" % (
+                os.path.expanduser('~'),
+                native_messenger_dirname,
+                "win_firefox_restart.ps1")
+
+            task_arg = "\"%s\"" % restart_ps1_path
+
+            open(restart_ps1_path, "w+").write(restart_ps1_content)
+
+            startupinfo = subprocess.STARTUPINFO()
+            startupinfo.dwFlags |= subprocess.STARTF_USESHOWWINDOW
+
+            subprocess.check_output(
+                ["powershell",
+                 "-NonInteractive",
```

Comment:
I added `-NoExit` here but all that seems to do is stop Firefox from exiting. The powershell window still flashes up unhelpfully.

Do you have any ideas for how I can debug this? I haven't touched profiledir: it's set to auto.

Reply:
@bovine3dom

#### ## Look into `win_firefox_restart.ps1`

Could you please paste the content of `$env:USERPROFILE\.tridactyl\win_firefox_restart.ps1` here? Check if you can run this script from a vanilla PowerShell prompt?

This is the script responsible to bring Firefox back. Note that this file is re-created every time `:restart` is invoked.

In PowerShell, the following command will copy the file into your clipboard. Please paste the content here to help debug further.

```Powershell
Get-Content "$env:USERPROFILE\.tridactyl\win_firefox_restart.ps1" | clip
```

#### ## Additional Debugging Steps

1. Keep Firefox (with the test profile) running, then in PowerShell run `&  "$env:USERPROFILE\.tridactyl\win_firefox_restart.ps1"`

2. Keep Firefox closed, then in PowerShell run `&  "$env:USERPROFILE\.tridactyl\win_firefox_restart.ps1"`

    - **[1]** Should put Powershell into a counting mode, waiting for the `parent.lock` file to be unlocked. Currently, this is set to `10 sec`, after that Powershell should show you an error message that Firefox needs to be restarted manually.

    - **[2]** Should immediately restart Firefox.

#### ## Background

Adding some more details below as reference. You might already know most of it. Hoping that someone else doing something similar might find it useful later on.

First of all, you'd find some verbose comments in `native_main.py::win_firefox_restart()` ([line 172][L172]) that both `Popen` family of functions exposed via `subprocess`, and `schtasks.exe` wasn't up to the task (at least for me on Windows). I still left them there, in case someone wants to take another stab at it using them later.

So, we end up doing the following _"orchestra"_ in JavaScript, Python and PowerShell:

1. User runs `:set profiledir [valid-profile-dir]`, followed by `:restart` in the JavaScript world.

   - JavaScript sends the following JSON object to the Python world:

```JSON
{
  "cmd":
        "win_firefox_restart",
  "profiledir":
        "C:\\Users\\<USERNAME>\\AppData\\Roaming\\Mozilla\\Firefox\\Profiles\\x5gxqprf.dev-edition-default"
}

```

2. Upon receiving the `win_firefox_restart` command, followed by some sanity checks, Python (`native_main.py`) creates `win_firefox_restart.ps1` in `~/.tridactyl/` ([line 236][L236])

3. Python then runs a blocking `subprocess.check_output()` call ([line 282][L282]) invoking another set of PowerShell commands to achieve the following:

   - Schedule a Windows task (analogous to Unix [Cron][cron]) to be run _only once_ in `delay` amount of seconds (currently set to `1 sec`), which will invoke `win_firefox_restart.ps1` from a completely separate process group, so that Firefox can be revived back even after its "death".

If steps [1] and [2] are successful, we send `code = 0` to the JavaScript world, at which point JavaScript gracefully calls `qall()` killing itself.

```Python
  reply = {
      "code": 0,
      "content": "Restarting in %d seconds..." % delay_sec
  }
```

In the meantime, if `delay = 1 sec` has passed, Windows task scheduler invokes `win_firefox_restart.ps1` (_briefly showing the PowerShell screen - I think you are currently getting stuck here_), which basically means that Windows is attempting to restart Firefox via the `win_firefox_restart.ps1` script. If `parent.lock` is blocked, the PowerShell screen should display the counting at this point.

One key element here to be noted is that quitting of Firefox and restarting are being done _asynchronously_ by separate parent processes (with `parent.lock` being the safeguard), and the `win_firefox_restart.ps1` embeds the critical restart functionality on Windows. 

Let me know if it helps. I'm happy to share more, or join on IRC if needed.

Thanks.


[L172]: https://github.com/gsbabil/tridactyl/blob/ecc451ae323f3e10f5c96b7eab90c328276ccfe5/native/native_main.py#L172

[L236]: https://github.com/gsbabil/tridactyl/blob/ecc451ae323f3e10f5c96b7eab90c328276ccfe5/native/native_main.py#L236

[L282]: https://github.com/gsbabil/tridactyl/blob/ecc451ae323f3e10f5c96b7eab90c328276ccfe5/native/native_main.py#L282

[cron]: https://en.wikipedia.org/wiki/Cron